### PR TITLE
Generic Guns rework

### LINKED ISSUE #35359: Update Generic Guns mod to be synchronized with current core content
# Describe the bug

The Generic Guns mod hasn't been updated in a couple of months -- and even then it might not have been complete as of that last update -- and guns have definitely been added since then and not incorporated.

# Steps To Reproduce

1. Start a new world with Generic Guns enabled
2. Observe errors (see below) due to outdated blacklists
3. Bring up debug menu (or look at the details in #34236 )
4. Search for `TAC-50` (or `tac50` in json)
5. Observe gun which is not defined in generic guns either as override or blacklist

# Expected behavior

Generic Guns, as an accessibility mod, should remain synchronized with the state of firearms, magazines, and ammunition in vanilla. Given that maintaining Generic Guns requires firearms knowledge, and anyone adding firearms must by definition possess that knowledge, I would argue it should be incumbent on contributors adding or updating guns to ensure Generic Guns is also updated.

# Screenshots

outdated blacklist definition 
image

Sum total of `tac50` references in master as of this writing:
image

# Versions and configuration

- OS: Linux
    - OS Version: Distributor ID: Ubuntu; Description: Ubuntu 19.10; Release: 19.10; Codename: eoan; 
- Game Version: 0.D-9180-gfcbe9b840b [64-bit]
- Graphics Version: Tiles
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Generic Guns [generic_guns]
]

# Additional information

Anyone working on this task might want to be aware of consistency check issues (see: #20320, found by way of  #31321 )

### PR DESCRIPTION
**Pull request by tenmillimaster**
#### Summary
```SUMMARY: Mods "Completely overhauls generic guns"```

#### Purpose of change
Generic Guns has been non-functional (to varying degrees) for a long, long time. It is difficult for content makers to maintain it, having to chase down multiple places for their work. This PR aims to close #35359 using the strategy outlined in #35427.

#### Describe the solution
Rather than having to copy and alter numerous items for the mod, Generic Guns will now utilize 'archetypal guns'(defined in #35483) that are defined in the mod, and are intended to stand on their own; thus, instead of carefully selecting which guns to black list, and which to not, **ALL MAINLINE** guns will get migrated instead to one of these archetypes.
In case  a gun doesn't adequately fit to one, a new archetype will be made.

The same approach is made for gunmods, magazines, ammunition, and recipes.

#### Describe alternatives you've considered
Just updating the blacklist.

#### Testing
Not much yet. I've loaded the game just prior, to make sure there aren't any load errors related to *this mod*. There is a feature this work is contingent on, which produces load errors, #35545
#### Additional context
This has been a HUGE project, and I have likely made mistakes. Any criticism, input, or suggestions are greatly appreciated.

As I was adding guns, I expanded the # of guns slightly to what I felt was a more complete list of archetypes.

### CODE REVIEW

**Comment by tenmillimaster** [comment]
Current list of known issues:
Pipe guns needed to get added to migrations.
Some mainline ammo is still craftable.
None of the Generic guns ammo is craftable?

**Comment by I-am-Erk** [comment]
So, I think we should make a CI check that makes sure all `type: ammo` and `type: guns` items are either migrated in Generic Guns, or are present on a whitelist to say they don't need migration.

File: data/mods/Generic_Guns/ammo/pistol_magnum.json [hunk 1]
```diff
@@ -0,0 +1,35 @@
+[
+  {
+    "id": "pistol_magnum_ball",
+    "type": "AMMO",
+    "copy-from": "44fmj",
+    "name": "magnum pistol ammo, ball",
+    "ammo_type": "ammo_pistol_magnum",
+    "casing": "pistol_magnum_casing",
+    "description": "Magnum pistol ammuniton, developed from early handgun hunting efforts and popularized by law enforcement usage.  It is more energetic than common pistol ammo.  'Ball' is an old term for full metal jacketed bullets which offer reduced fouling, better reliability, and slightly improved penetration compared to plain lead."
```
Comment (I-am-Erk) [inline data/mods/Generic_Guns/ammo/pistol_magnum.json]: You said "energetic" here and "powerful" below. It's not a bit deal but you could go for consistency.
```suggestion
    "description": "Magnum pistol ammuniton, developed from early handgun hunting efforts and popularized by law enforcement usage.  It is more powerful than common pistol ammo.  'Ball' is an old term for full metal jacketed bullets which offer reduced fouling, better reliability, and slightly improved penetration compared to plain lead."
```

File: data/mods/Generic_Guns/firearms/black_powder.json [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    "description": "This antiquated single shot pistol would complete a pirate's outfit nicely.  While its loading may be slow, it does not require brass casings to fire, and is theoretically sustainable.  Range and accuracy are hampered by lack of rifling, but this old design is still plenty lethal.",
+    "clip_size": 1
+  },
+  {
+    "id": "black_powder_revolver",
+    "copy-from": "colt_army",
+    "type": "GUN",
+    "name": "antique revolver",
+    "ammo": "ammo_black_powder",
+    "description": "This is an old revolver designed during the period of westward expansion. Black powder paper cartridges must be loaded for each shot, which is a fairly lengthy process.   Despite its age, this would perform adequately against most two-legged threats. ",
```
Comment (I-am-Erk) [inline data/mods/Generic_Guns/firearms/black_powder.json]:
```suggestion
    "description": "This is an old revolver designed during the period of westward expansion.  Black powder paper cartridges must be loaded for each shot, which is a fairly lengthy process.   Despite its age, this type of weapon would perform adequately against most two-legged threats.",
```

File: data/mods/Generic_Guns/firearms/pistol.json [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    "description": "Since the early days of western expansion, when ammunition supply was infrequent, shooters sought to share ammo between their sidearms and long arms.  This tube magazine lever action carbine allows the shooter to reduce the number of calibers carried and squeeze more power from their pistol ammunition.",
+    "clip_size": 10
+  },
+  {
+    "id": "pistol_machine",
+    "copy-from": "glock_18c",
+    "type": "GUN",
+    "name": "machine pistol",
+    "ammo": "ammo_pistol",
+    "description": "This pistol is a tiny machinegun you can stuff into a holster, with which you could dump its magazine at a blistering rate into any close range foes.  Machine pistols mostly see use by vehicle crewmen or bodygaurds of VIPs. Due to its preposterous rate of fire it is difficult to fire controllably.",
```
Comment (I-am-Erk) [inline data/mods/Generic_Guns/firearms/pistol.json]:
```suggestion
    "description": "This pistol is a tiny machinegun you can stuff into a holster, with which you could dump its magazine at a blistering rate into any close range foes.  Machine pistols mostly see use by vehicle crewmen or bodygaurds of VIPs.  Due to its preposterous rate of fire it is difficult to control.",
```

File: data/mods/Generic_Guns/firearms/pistol.json [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    "description": "These small carbines share ammunition and magazines with common pistols, offering a more controllable carbine than a normal rifle, while also reducing ammunition costs.  Because of their compatibility, they pair well with duty pistols, allowing one to transition to a more stable weapon without carrying extra ammo or magazines.",
+    "magazines": [ [ "ammo_pistol", [ "pistol_mag",  "pistol_smg_mag" ] ] ]
+  },
+  {
+    "id": "pistol_revolver",
+    "copy-from": "ruger_lcr_38",
+    "type": "GUN",
+    "name": "police revolver",
+    "ammo": "ammo_pistol",
+    "description": "This may not be an actual police firearm, but similar revolvers chambered for standard defensive calibers like this were a favorite of police departments for nearly a century, up until the 1986 Miami shootout.  Afterwards, the slow reloading and shooting of revolvers were considered liabilities; still, this model's accuracy and moderate recoil make for a serviceable sidearm, and there are no magazines for you to lose or damage.",
```
Comment (I-am-Erk) [inline data/mods/Generic_Guns/firearms/pistol.json]:
```suggestion
    "description": "Revolvers like this, chambered for standard defensive calibers, were a favorite of police departments for nearly a century, up until the 1986 Miami shootout.  Afterwards, the slow reloading and shooting of revolvers were considered liabilities; still, this model's accuracy and moderate recoil make for a serviceable sidearm, and there are no magazines for you to lose or damage.",
```
YMMV but I think that's more concise

File: data/mods/Generic_Guns/firearms/rifle.json [hunk 5]
```diff
@@ -0,0 +1,56 @@
+[
+  {
+    "id": "rifle_assault",
+    "copy-from": "ar10",
+    "type": "GUN",
+    "name": "service rifle",
+    "ammo": "ammo_rifle",
+    "description": "Borne from years of refinement .  The products of decades of improvement, rifle such as this are handy, reliable, and adaptable.   Short of large creatures and light vehicles, this should take care of most of your problems out to several hundred meters.",
```
Comment (I-am-Erk) [inline data/mods/Generic_Guns/firearms/rifle.json]:
```suggestion
	    "description": "Borne from years of refinement.  The products of decades of improvement, rifle such as this are handy, reliable, and adaptable.   Short of large creatures and light vehicles, this should take care of most of your problems out to several hundred meters.",
```

Comment (tenmillimaster) [reply data/mods/Generic_Guns/firearms/rifle.json]: @I-am-Erk oh right, I never finished this description...

File: data/mods/Generic_Guns/ammo/casings.json [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    "type": "GENERIC",
+    "name": "rifle casing",
+    "description": "An empty casing from a rifle round."
+  },
+  {
+    "id": "rifle_huge_casing",
+    "copy-from": "50_casing",
+    "type": "GENERIC",
+    "name": "huge rifle casing",
+    "description": "A huge casing from an huge rifle round."
```
[the remark this answers is not in the document]
Comment (tenmillimaster) [reply data/mods/Generic_Guns/ammo/casings.json]: Poor attempt at humor.

File: data/mods/Generic_Guns/firearms/rifle.json (hunk 5 above)
Comment (Brian-Otten) [reply data/mods/Generic_Guns/firearms/rifle.json]: Just went ingame to check all the itemgroups. going to note down items that need migrating

luty smg .40
luty smg .45
luty smg 9mm
winchester m37
Browning auto 5
CZ-75
Glock 31
hi-point c9
Kel tec p32
Kel-tec p3AT
kel-tec pf-9
marlin 1895
mossberg 930 spx
PT 603
remington 1100 competition
remington 700
remington 870 express
ruger LCP
Sig p320
Walther p22
Walter p38
AT 4
m74 rocket clip
m110A1
Glock 18c
M17
m1911A1
Mk 23 MOD
mcmillan tac-50
mossberg 590A
remington 870 MCS
Magnum research BFR
Saiga-410
Taurus raging judge magnum
USAS 12
Uberti cattleman
Browning hi-power .40
Walther CCP
Walther PPQ
S&W model 10
Walther ppq 9mm
m45A1
Bond arms derringer
Ar-15
Hk417
1874 sharps
1887 bootleg shotgun
Colt lighting .45 carbine
m1987 trench gun
mac-11
Tavor TS12
IWI tavor x95
ar-10
Marlin 1895 SBL
Remington 700
Remington ACR .300
Cobray street sweeper
Franchi spas-12
PTR 603
H&K mp5k-pdw
H&K mpsd
.45-70 LFN
usas-12 drum magazine
saiga-410 drum magazine
usas-12 box magazine
saiga-410 box magazine
AF2011A1 magazine
.300ac blackout conversion kit
.410 000


ammo_shotgun_obscure spawns nothing
guns_launcher_grenade spawns nothing
guns_launcher_improvised spawns nothing
guns_smg_obscure spawns nothing
mags_shotgun_common spawns nothing

That's all the ones i could find easily, there might still be more.

**Comment by Brian-Otten** [comment]
Just went ingame to check all the itemgroups. going to note down items that need migrating

luty smg .40
luty smg .45
luty smg 9mm
winchester m37
Browning auto 5
CZ-75
Glock 31
hi-point c9
Kel tec p32
Kel-tec p3AT
kel-tec pf-9
marlin 1895
mossberg 930 spx
PT 603
remington 1100 competition
remington 700
remington 870 express
ruger LCP
Sig p320
Walther p22
Walter p38
AT 4
m74 rocket clip
m110A1
Glock 18c
M17
m1911A1
Mk 23 MOD
mcmillan tac-50
mossberg 590A
remington 870 MCS
Magnum research BFR
Saiga-410
Taurus raging judge magnum
USAS 12
Uberti cattleman
Browning hi-power .40
Walther CCP
Walther PPQ
S&W model 10
Walther ppq 9mm
m45A1
Bond arms derringer
Ar-15
Hk417
1874 sharps
1887 bootleg shotgun
Colt lighting .45 carbine
m1987 trench gun
mac-11
Tavor TS12
IWI tavor x95
ar-10
Marlin 1895 SBL
Remington 700
Remington ACR .300
Cobray street sweeper
Franchi spas-12
PTR 603
H&K mp5k-pdw
H&K mpsd
.45-70 LFN
usas-12 drum magazine
saiga-410 drum magazine
usas-12 box magazine
saiga-410 box magazine
AF2011A1 magazine
.300ac blackout conversion kit
.410 000

ammo_shotgun_obscure spawns nothing
guns_launcher_grenade spawns nothing
guns_launcher_improvised spawns nothing
guns_smg_obscure spawns nothing
mags_shotgun_common spawns nothing

That's all the ones i could find easily, there might still be more.

**Comment by esotericist** [comment]
> This apparently didn't work because the test itemgroup function in the debug menu does not properly take migrations into account. So this list is sadly mostly invalid.

migrations currently only happen on save game load, because that mechanism was originally only intended to be used for obsoleted items (and thus things that would no longer be in item lists). 

engineering is needed for this to work right.

See: #35545

**Comment by tenmillimaster** [comment]
I have been playtesting somewhat with a compiled copy of #35545. Most everything seems to be in order, I missed a few migrations.

Currently I'm having issues with magazines- https://media.discordapp.net/attachments/598529174302490644/666523696155852800/unknown.png?width=349&height=197

they have a copy-from from a similar mainline mag, and I was under the impression redefining a field would override it for the new item. This wasn't happening- I tried "delete" : { ammo_type: "old magammotype"} as well, but that hasn't worked.

**Comment by tenmillimaster** [comment]
I've gone and cleaned up some more of the recipes, threw in a very hacky hack to make vehicle turrets and their ammo spawns happy, for 0.E (until migration works properly with vehicle ammo).  There was a circular dependency error caused by 'bp_shot_scrap' and a recipe copying from it, related to the order in which they were defined, so I've gone and fixed that too.

I've added black powder versions of all the ammo, and de-referenced most of the recipes since copying and obsoleting wasn't going to work.

I've split all the pipe-weapons (pipe on a stick with a hammer) apart from the repeaters (somewhat well made firearms) in recipes, and have attempted to rationalize their difficulties and components.

**Comment by I-am-Erk** [comment]
Well let's see how that goes

**Pull request merged**